### PR TITLE
Added sort_by_label function for sorting by label values.

### PR DESCRIPTION
**Pull request by pstibrany**
Added sort_by_label(vector, labelName) function for sorting by label values. If such function is welcome, I can write documentation for it.

---

This change is [Reviewable](https://reviewable.io/reviews/prometheus/prometheus/1533)

### CODE REVIEW

File: promql/functions.go [hunk 1]
```diff
@@ -298,6 +299,14 @@ func funcSortDesc(ev *evaluator, args Expressions) model.Value {
 	return vector(byValueSorter)
 }
 
+// === sort_by_label(node model.ValVector, model.ValString) model.ValVector ===
+func funcSortByLabel(ev *evaluator, args Expressions) model.Value {
+	l := ev.evalString(args[1])
+	s := vectorByLabelHeap{vector: ev.evalVector(args[0]), labelName: l.String()}
+	sort.Sort(s)
```
Comment (brian-brazil) [inline promql/functions.go]: This should use a stable sort, so that they can be chained

Comment (pstibrany) [inline promql/functions.go]: Thank you for looking at my change. I don't quite follow you though. How is this different from funcSort or funcSortDesc?

Comment (brian-brazil) [inline promql/functions.go]: Unlike values, a timeseries can have multiple labels. If I want to sort by A, then B I can do `sort_by_label(sort_by_label(my_vector, "B") "A")` but only if the sort is stable.

Comment (pstibrany) [inline promql/functions.go]: Ah, ok. Thanks for explanation. I'll make sure it uses stable sort.

**Comment by grobie** [comment]
What are the use cases for this outside of the console view in `/graph`? If that's the use case, I'd rather implement it as javascript UI feature than as a language addition.

**Comment by pstibrany** [comment]
That is indeed my primary use case.

**Comment by RichiH** [comment]
@grobie The problem is that you don't know where else this might be consumed. I think there's a valid case for sorting before handling it in Grafana.

Implementing this anywhere else means it needs to be implemented several times.

**Comment by grobie** [comment]
But where do we stop? The next request will be a `sort_by_label_desc` (which is an awful function name) and with the the same argumentation we'll need to many more UI related functions.

I'd rather implement UI/presentation related features twice in graph consoles and grafana than cluttering PromQL and fighting an endless battle about what to add and what not "sort_by_label is already part of the language, there my alias() function makes sense as well".

File: promql/functions.go (hunk 1 above)
Comment (fabxc) [inline promql/functions.go]: Sounds like a good use case for variadic arguments.

**Comment by LuboVarga** [comment]
Suggestion 1:
What about to add this feature (sort_by_label) into prometheus/graph screen, where query will be modified and sorted in html (javascript)? So we can use this query: sort_by_label(up, "job")` or `sort_by_label(up, "instance")` without searching for particular line in result table after each execution? (I know, I can use also `up{job="JobWhichIAmInterestedRightNow"}` query, but often there are a bit more services and writing query to match all is not very comfortable).

Suggestion 2:
What about to add "sort" function to resulting table in prometheuse graph/console? Just familiar sort thing when user clicks on column name. Simple, but I hope doable without any objections. This would address my ad-hoc use cases.

PS: Probably directly graphanas UI code can be reused for this.
PS2: Implementing sort in UI will make limitation to be able to use it only on top level.

**Comment by 123BLiN** [comment]
but what is the status of this PR?)

**Comment by RichiH** [comment]
Looking at it, the feature is somewhat contentious, but still not rejected outright. After these two years, I also think it would be better to keep things outside of core PromQL, but I am wondering if it makes sense to have a second layer behind PromQL proper which allows to format, sort, and mangle data.

As goes beyond this PR, I suggest carrying this discussion over to the ML or to an issue.

**Comment by josephine-rueckert** [comment]
@grobie Sorting series is not supported by Grafana. From what I read in the following issues, it seems like they display the series in the same order they get it from Prometheus, thus expecting the query to provide a sorting feature.

>The order in which series are drawn is the same as the order returned by your Prometheus query, you need to sort your series in your query.
https://github.com/grafana/grafana/issues/13325

**Comment by zkdev** [comment]
So there is no sorting layer within PromQL and Grafana says sorting has to be done inside the query.
Quite unfortunate situation - will there be an update regarding this topic? @RichiH

**Comment by HuanGong** [comment]
i also need this feature, any plan about this PR?

**Comment by jacksontj** [comment]
+1 on havin this, sounds like the PR itself is dead, is prometheus open to someone re-opening this PR and getting it in?

**Comment by nhartner** [comment]
What kind of query language doesn't support sorting? Why is there such ambivalence by Prometheus maintainers towards a basic feature of nearly every query language. SQL supports sorting. GraphQL supports sorting. LINQ supports sorting. PromQL is definitely in the minority here.

**Comment by f1-outsourcing** [comment]
I just wanted to use this. Is just helpful in verifying query output.

**Comment by Powie** [comment]
+1

**Comment by RichiH** [comment]
FYI, I put this wider topic on the dev summit agenda under "Extending PromQL". I still think this feature has merit and I think the time is ripe to revisit older discussions within the project.

Side note: Adding `+1` comments does not help the discussion at large, it's largely noise; this is literally why GitHub implemented reactions to comments.

**Comment by trallnag** [comment]
Disclaimer: **Not** a shill. MetricsQL from VictoriaMetrics supports sorting by label value.

```
sort_by_label(q, label) and sort_by_label_desc(q, label) for sorting time series by the given label.
```

https://github.com/VictoriaMetrics/VictoriaMetrics/wiki/MetricsQL

**Comment by ogxd** [comment]
I think this feature is still requested a lot. I think you are right @grobie that it would require sort_by_label_desc as well, but I don't think we'll need to "add many more UI related functions" then (or do you an example?). Also, all functions are "UI related" since the end goal is always to visualize at some point. 
Could we consider reopening the request?

**Comment by roidelapluie** [comment]
Yes, I think it is reasonable to have this feature. Reopening the pull request is probably not with it because of all the rebase needed.

**Comment by RichardoC** [comment]
Is there an alive feature for this? All the references I can find have been closed.

**Comment by galexrt** [comment]
I have just stumbled upon this as an issue for myself as well and went ahead updated the code to work with the latest code base (PR #11299) to hopefully "re-ignite" discussions for this functionality.

I'll look into joining one of the upcoming Prometheus Contributor Office hours to see how we can discuss this PromQL functionality.

**Comment by pixelrebel** [comment]
IMO, this implementation is the datasource's responsibility.  The function names should use terminology that relates to its platform.  There is no concept of a label in most other datasources.  So `sort_by_label` should be in PromQL, in the same way `ORDER BY` should be in SQL.  

Without this feature, we can not have reliable visualizations.  For example, I would like to use the "Config by query results" transformation in grafana to display my gauges with real numbers instead of percentages.  Suppose I have two disks labeled `disk1` and `disk2`.  Suppose their used/capacity ratios are 50/100 and 5/20 respectively.  In order to display a gauge with capacity as the max threshold, the order for the used queries and the capacity queries needs to line up by label.  But it's totally random.   So sometimes I get 50/100 and 5/20.  Other times I get 50/20 and 5/100.  The only way around this is to use a single percentage query, but that is not a helpful visualization for some cases.  

`sort_by_label` should be fundamental to the query language.  I'm surprised this is even debatable.

**Pull request closed without merging**